design/ceph-deprecate-old-osd-types.md

```diff
@@ -0,0 +1,212 @@
+# Deprecating old Ceph OSD types (May 2019)
+**Targeted for v1.1-v1.2**
+
+## Background
+Rook's current implementation for preparing and starting OSDs includes several different patterns
+that have evolved over Rook's lifetime. These patterns include different code paths for supporting
+directories and devices as storage targets, and further separated into support for both Filestore
+and Bluestore as backing stores. There is further complication added by support for `ceph-volume` as
+a tool for storage provisioning versus continued support for legacy installations of Rook which used
+manual partitioning.
+
+There has been talk for quite some time about removing support for directories as storage targets,
+as this is a use case in Ceph that is not supported for production. Additionally, Filestore is
+effectively in sustaining mode, and Bluestore is the defacto replacement for it.
+
+All of the various OSD support patterns/code-paths need changes to allow the [planned updates to
+Rook's configuration strategy](/design/ceph-config-updates.md) to proceed. The various code paths
+are intertangled and difficult to separate, and this presents a huge hurdle for the config changes.
+In general, the intertangled OSD code poses a risk for long-term maintenance, as it may be more
+prone to bugs being introduced, and it has and will continue to slow development.
+
+In light of this background information, this design proposes three high-level goals:
+1. Identify OSD features/code-paths that should be deprecated due to having become outmoded in some form
+2. Propose a new architecture for Rook's OSD code that is compatible with the planned config updates
+and that is extensible for the future.
+3. Provide a timeline for deprecating outmoded, and provide users tools for transitioning their
+   deployments which use old methods to begin using the new methods.
+
+
+## Deprecated features and code paths
+This design proposes to deprecate ...
+- the use of directories as storage targets
+- support for the Filestore backing store
+- support for devices not provisioned with `ceph-volume`
+
+
+## New OSD operator architecture
+The new OSD operator design focuses only on devices provisioned with Bluestore using `ceph-volume`.
+Some design thought is given to allowing space for support of post-Bluestore backing stores.
```

Comment:
Another consideration is for creating OSDs on top of PVCs as described in #919, which is planned for 1.1. I'm not sure it will affect this design change, but just something to be aware of.

Reply:
I will have to consider this more. It's clearly a feature that users want, so I think it should go in, but I have this sinking feeling that users are going to create some very degenerate Ceph clusters with this feature, and we are going to have to explain why we can only support that to a certain point. We're already seeing one user wanting to create a Ceph cluster using a single block device on each node and then resize the underlying block device, which is absolutely bonkers from Ceph's standpoint being a distributed object store designed for real-world hardware.

With the scattered thoughts I have currently, I think the PVC-backed use case will have to be a feature of Rook-managed mode only.

Reply:
With the way K8s is moving, the block PVs will help us more than just the scenarios users are asking for in #919. It helps us move into a world where K8s is aware of the underlying storage and can manage the raw devices for us. While the devil is in the details, I would expect this will also help with our permissions challenge with mounting /dev.

Reply:
In the update, I broke Rook-managed mode into 2 submodes: an automatic mode and a declarative mode, and the PVC use case is applicable only for the declarative mode.